services/motion/builtin/builtin.go

```diff
@@ -537,3 +538,73 @@ func (ms *builtIn) planMoveOnMap(
 	plan, err := motionplan.FrameStepsFromRobotPath(f.Name(), solutionMap)
 	return plan, kb, err
 }
+
+// ErrorState will provide a pose P that is the kinematic base's offset from where it would be if it were to follow a path perfectly.
+// This pose P is such that spatialmath.Compose(nominalPose, P) will yield the actual pose seen by the localizer.
+func ErrorState(
+	ctx context.Context,
+	kb kinematicbase.KinematicBase,
+	plan [][]referenceframe.Input,
+	currentNode int,
+) (spatialmath.Pose, error) {
+	if currentNode < 0 || currentNode >= len(plan) {
+		return nil, fmt.Errorf("cannot get errorState for node %d, must be >= 0 and less than plan length %d", currentNode, len(plan))
+	}
+
+	// Get pose of the base via its localizer. The offset between the localizer and its base should already be accounted for.
+	actualPose, err := kb.CurrentPosition(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var nominalPose spatialmath.Pose
+	frame := kb.Kinematics()
+	if _, ok := frame.(tpspace.PTGProvider); ok {
+		// TODO: The `rectifyTPspacePath` in motionplan does basically this. Deduplicate.
+		runningPose := spatialmath.NewZeroPose()
+		for i, wp := range plan {
+			wpPose, err := frame.Transform(wp)
+			if err != nil {
+				return nil, err
+			}
+			runningPose = spatialmath.Compose(runningPose, wpPose)
+			if i == currentNode {
+				break
+			}
+		}
+
+		// Determine how far through the current trajectory we are
+		currentInputs, err := kb.CurrentInputs(ctx)
+		if err != nil {
+			return nil, err
+		}
+		currPose, err := frame.Transform(currentInputs)
+		if err != nil {
+			return nil, err
+		}
+
+		nominalPose = spatialmath.Compose(runningPose, currPose)
+	} else {
+		if len(plan) < 2 {
+			return nil, errors.New("Diff drive motion plan must have at least two waypoints")
+		}
+		nominalPose, err = frame.Transform(plan[currentNode])
+		if err != nil {
+			return nil, err
+		}
+		if currentNode > 0 {
+			pastPose, err := frame.Transform(plan[currentNode-1])
+			if err != nil {
+				return nil, err
+			}
+			// diff drive bases don't have a notion of "distance along the trajectory between waypoints", so instead we compare to the
+			// nearest point on the straight line path.
```

Comment:
Could we use a metric here instead of doing this spatialmath?

Reply:
A metric returns a float as a distance value, not a full pose